wallet/datastore.go

```diff
@@ -0,0 +1,333 @@
+	// NOTE on conflict do nothing because none of the wallet information is updateable
+	statement := `
+	INSERT INTO wallets (id, provider, provider_id, public_key)
+	VALUES ($1, $2, $3, $4)
+	ON CONFLICT DO NOTHING`
+	_, err := pg.RawDB().Exec(
+		statement,
+		wallet.ID,
+		wallet.Provider,
+		wallet.ProviderID,
+		wallet.PublicKey,
+	)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// SetAnonymousAddress sets the anon addresses of the provided wallets
+func (pg *Postgres) SetAnonymousAddress(ID string, anonymousAddress *uuid.UUID) error {
+	statement := `
+	UPDATE wallets
+	SET
+			anonymous_address = $2
+	WHERE id = $1
+	`
+	_, err := pg.RawDB().Exec(
+		statement,
+		ID,
+		anonymousAddress,
+	)
+	return err
+}
+
+// LinkWallet links a wallet and transfers funds to newly linked wallet
+func (service *Service) LinkWallet(
+	ctx context.Context,
+	info *walletutils.Info,
+	transaction string,
+	anonymousAddress *uuid.UUID,
+) error {
+	// do not confirm this transaction yet
+	tx, err := service.SubmitCommitableAnonCardTransaction(
+		ctx,
+		info,
+		transaction,
+		"",
+		false,
+	)
+	if err != nil {
+		return handlers.WrapError(err, "unable to verify transaction", http.StatusBadRequest)
+	}
+	if tx.UserID == "" {
+		err := errors.New("user id not provided")
+		return handlers.WrapError(err, "unable to link wallet", http.StatusBadRequest)
+	}
+	providerLinkingID := uuid.NewV5(walletClaimNamespace, tx.UserID)
+	if info.ProviderLinkingID != nil {
+		// check if the member matches the associated member
+		if !uuid.Equal(*info.ProviderLinkingID, providerLinkingID) {
+			return handlers.WrapError(errors.New("wallets do not match"), "unable to match wallets", http.StatusForbidden)
+		}
+		if anonymousAddress != nil && info.AnonymousAddress != nil && !uuid.Equal(*anonymousAddress, *info.AnonymousAddress) {
+			err := service.Datastore.SetAnonymousAddress(info.ID, anonymousAddress)
+			if err != nil {
+				return handlers.WrapError(err, "unable to set anonymous address", http.StatusInternalServerError)
+			}
+		}
+	} else {
+		err := service.Datastore.LinkWallet(info.ID, providerLinkingID, anonymousAddress)
+		if err != nil {
+			status := http.StatusInternalServerError
+			if err == ErrTooManyCardsLinked {
+				status = http.StatusConflict
+			}
+			return handlers.WrapError(err, "unable to link wallets", status)
+		}
+	}
+
+	if decimal.NewFromFloat(0).LessThan(tx.Probi) {
+		_, err := service.SubmitCommitableAnonCardTransaction(ctx, info, transaction, "", true)
+		if err != nil {
+			return handlers.WrapError(err, "unable to transfer tokens", http.StatusBadRequest)
+		}
+	}
+	return nil
+}
+
+// TxLinkWalletInfo pass a tx to set the anonymous address
+func (pg *Postgres) TxLinkWalletInfo(tx *sqlx.Tx, ID string, providerLinkingID uuid.UUID, anonymousAddress *uuid.UUID) error {
+	statement := `
+	UPDATE wallets
+	SET
+			provider_linking_id = $2,
+			anonymous_address = $3
+	WHERE id = $1;`
+	_, err := tx.Exec(
```

Comment:
Should we check how many rows we've modified so that this was updated?

Reply:
:+1:

Reply:
https://github.com/brave-intl/bat-go/pull/314/commits/d9ad61f5ecbbf6e774a35d6481720162bacc3f87